### PR TITLE
fix(router): keep detached route subtree contexts isolated and intact

### LINKED ISSUE #57285: Child routes don't always render on reused parent route
### Which @angular/* package(s) are the source of the bug?

router

### Is this a regression?

No

### Description

I have a basic app in the following structure:
```
['/', 'events', '', ':id', 'edit']
['/', 'chats', ':id', 'edit']
```
I have a custom route reuse strategy to reuse `/events` and `/events/:id` as well as `/chats` and `/chats/:id`.

In general, this works. However, navigating to the child route 'edit' (or any other child routes if there were any) will not always work.
It seems to depend on how deep the child route is.
In my example, `/events/:id/edit` will break, but `/chats/:id/edit` will not.
Note how `events` has a path equal to `''`. 

What should happen:
1. Open/refresh the demo.
2. Select Chat 1.
3. Click Edit.
4. The edit box shows up as expected.
5. Once you have done this, the entire app will continue working as expected, clicking through the different sections.

To break editing:
1. Open/refresh the demo (need to open to the `/` or `/events`)
2. Click Chats
3. Click Events
4. Click Event 1
5. Click Edit.
6. The edit box will not show up.
7. It will not be possible to edit an event until you refresh the page.

There are other paths to break it. The key thing to break it is that the `/events` page needs to be reused, before loading one of its child routes.

Works:
Start from Chats -> Events -> Event 1 -> Edit item
Start from Chats -> Events -> Event 1 -> Chats -> Events -> Event 1 -> Edit item

Doesn't work:
Start from Chats -> Events -> Chats -> Events -> Event 1 -> Edit item

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/broken-reuse-child

### Please provide the exception or error you saw

```true
Child route is not rendered into <router-outlet></router-outlet>
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 18.1.1
Node: 18.20.3
Package Manager: npm 10.2.3
OS: linux x64

Angular: 18.1.1
... cli, common, compiler, compiler-cli, core, platform-browser
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1801.1
@angular-devkit/build-angular   18.1.1
@angular-devkit/core            18.1.1
@angular-devkit/schematics      18.1.1
@schematics/angular             18.1.1
rxjs                            7.8.1
typescript                      5.4.5
zone.js                         0.14.8
```

### Anything else?

The correct title will still appear in the browser.
There are no console errors.

### PR DESCRIPTION
**Pull request by arturovt**
When a route is detached for `RouteReuseStrategy`, its component and child
`RouterOutlet`s stay alive and keep referencing the `ChildrenOutletContexts`
they were created with. `detachAndStoreRouteSubtree` used to call
`onOutletDeactivated()`, which swaps that object's Map for an empty one, so
after re-attaching, the inner outlet and the router read from two different
context trees and deeper child routes (e.g. an `edit` route under a reused
list) never rendered.

Now, on detach: take the child contexts map as-is (destruction still calls
`onOutletDeactivated()` to prune) and give the `OutletContext` a fresh
`ChildrenOutletContexts`. This keeps the detached component rendering from
its stored map once re-attached, and prevents whatever activates next in
the same parent outlet (e.g. a sibling tab) from mutating or wiping that
stored map.

Fixes https://github.com/angular/angular/issues/57285

### CODE REVIEW

**Comment by arturovt** [comment]
@atscott friendly :ping_pong:

File: packages/router/src/router_outlet_context.ts [hunk 1]
```diff
@@ -68,9 +68,13 @@ export class ChildrenOutletContexts {
...
-    const contexts = this.contexts;
-    this.contexts = new Map();
-    return contexts;
+    // Return this Map as-is instead of replacing it with a new one. A `RouterOutlet` that lives
+    // inside a detached (stored) component keeps a reference to this same `ChildrenOutletContexts`
+    // instance. If we swapped in a fresh Map, that outlet would keep using the old one while the
+    // router uses the new one once the route is re-attached, so deeper child routes would fail to
+    // render (see #57285). Contexts for outlets that were really destroyed are removed separately
+    // by `onChildOutletDestroyed()`.
+    return this.contexts;
```
Comment (atscott) [inline packages/router/src/router_outlet_context.ts]: AGENT: Returning `this.contexts` without clearing it causes a memory leak when routes are destroyed without `RouteReuseStrategy`.

In [`activate_routes.ts`](https://github.com/angular/angular/blob/main/packages/router/src/operators/activate_routes.ts#L134), `deactivateRouteAndOutlet()` calls `context.children.onOutletDeactivated()` specifically to clean up the child context tree:
```ts
if (context.outlet) {
  // Destroy the component
  context.outlet.deactivate();
  // Destroy the contexts for all the outlets that were in the component
  context.children.onOutletDeactivated();
}
```

Because `ChildrenOutletContexts` is a root-level singleton, if `this.contexts` is never cleared on deactivation, every destroyed route's nested `OutletContext` -> `ChildrenOutletContexts` tree remains rooted in memory and cannot be garbage collected.

Route detachment and route destruction have different requirements:
- **Detaching (`detachAndStoreRouteSubtree`)**: The component and child outlets remain alive, so its context map should be preserved intact.
- **Destroying (`deactivateRouteAndOutlet`)**: The component is destroyed, so child contexts should be cleared to prune the dead tree.

`onOutletDeactivated()` should continue clearing `this.contexts` for the destruction path. In `detachAndStoreRouteSubtree`, we can instead mark `contexts` as `/** @internal */` and read `context.children.contexts` directly without wiping the map.

**Review by atscott — CHANGES_REQUESTED** [review]
AGENT: Making `onOutletDeactivated()` unconditionally retain `this.contexts` introduces a memory leak during normal route destruction.

`ChildrenOutletContexts` is a root-level singleton (`@Injectable({providedIn: 'root'})`). When a route is destroyed without a `RouteReuseStrategy`, [`deactivateRouteAndOutlet`](https://github.com/angular/angular/blob/main/packages/router/src/operators/activate_routes.ts#L134) calls `context.children.onOutletDeactivated()` to prune the child `OutletContext` subtree so it can be garbage collected. If `onOutletDeactivated()` stops clearing `this.contexts`, the entire nested tree of `OutletContext` and `ChildrenOutletContexts` instances remains retained in memory for all routes navigated across the session.

Route detachment and route destruction should be handled separately:
1. **Destruction (`deactivateRouteAndOutlet`)**: Keep resetting `this.contexts = new Map()` on `onOutletDeactivated()` so destroyed subtrees are pruned.
2. **Detachment (`detachAndStoreRouteSubtree`)**: When detaching a route, do not call `onOutletDeactivated()`. Instead, read `context.children.contexts` directly (making the property `/** @internal */`) so the map is preserved on the detached component without clearing it.

File: packages/router/src/operators/activate_routes.ts [hunk 2]
```diff
@@ -107,7 +107,11 @@ export class ActivateRoutes {
 
     if (context && context.outlet) {
       const componentRef = context.outlet.detach();
-      const contexts = context.children.onOutletDeactivated();
+      // The detached component and its child `RouterOutlet`s stay alive and keep referencing this
```
Comment (atscott) [inline packages/router/src/operators/activate_routes.ts]: AGENT: If another component is activated in this same parent outlet while this route is detached (e.g. navigating `/tabs/tab1` -> `/tabs/tab2`), that new component will receive the same `context.children` instance, mutating and eventually wiping `contexts` upon deactivation.

Reset `context.children` to a fresh instance upon detach to isolate the detached component's context tree:
```ts
if (context && context.outlet) {
  const componentRef = context.outlet.detach();
  const contexts = context.children.contexts;
  context.children = new ChildrenOutletContexts(context.injector);
  this.routeReuseStrategy.store(route.value.snapshot, {componentRef, route, contexts});
}
```

**Review by atscott — CHANGES_REQUESTED** [review]
AGENT: There is one remaining edge case: if another component is activated in the same parent outlet while a route is detached (e.g., tab switching `/tabs/tab1` -> `/tabs/tab2` where the parent `/tabs` outlet stays alive):

1. When `tab1` is detached, `context.children` is left on the parent `OutletContext`.
2. When `tab2` activates in that same outlet, it is given that same `context.children` instance, so its child outlets overwrite `tab1`'s entries in `contexts`.
3. When `tab2` is later destroyed, `deactivateRouteAndOutlet()` calls `context.children.onOutletDeactivated()`, wiping `this.contexts = new Map()` on `tab1`'s stored map.
4. Reattaching `tab1` then fails to render its deeper child routes.

To fix this, reset `context.children` to a fresh `ChildrenOutletContexts` instance when detaching in `detachAndStoreRouteSubtree`:
```ts
if (context && context.outlet) {
  const componentRef = context.outlet.detach();
  const contexts = context.children.contexts;
  context.children = new ChildrenOutletContexts(context.injector);
  this.routeReuseStrategy.store(route.value.snapshot, {componentRef, route, contexts});
}
```

Please also add a test covering this sibling tab-switch scenario alongside the existing shell-recreation test.

---
*Note: `RouteReuseStrategy` is still fundamentally broken for certain scenarios (such as re-parenting detached components across different ancestor component/injector hierarchies) and cannot be fully fixed because Angular's `ElementInjector` tree is static and cannot be re-wired to a new parent at runtime.*

File: packages/router/src/operators/activate_routes.ts [hunk 3]
```diff
@@ -107,7 +107,14 @@ export class ActivateRoutes {
...
+      // `ChildrenOutletContexts` instance. Take its map as-is (unlike the destruction path, which
+      // calls `onOutletDeactivated()` to prune) and give the `OutletContext` a fresh instance, so:
+      //   - the detached component keeps rendering from the stored map once re-attached (#57285);
+      //   - whatever activates next in this same parent outlet (e.g. a sibling tab) can't mutate
+      //     or wipe the detached component's stored map.
+      const contexts = context.children.contexts;
+      context.children = new ChildrenOutletContexts(context.injector);
```
Comment (atscott) [inline packages/router/src/operators/activate_routes.ts]: AGENT: `context.injector` is a getter that evaluates to `this.route?.snapshot._environmentInjector ?? this.rootInjector`.

At this point, `context.route` still points to the detached route. If that route has route-level `providers: [...]` or is lazy-loaded, `context.injector` will pass that route's local `_environmentInjector` as the `rootInjector` to the new `ChildrenOutletContexts`.

Instead, `OutletContext` should re-initialize `this.children` with its own `this.rootInjector` (via an internal `resetChildren()` method):
```ts
context.resetChildren();
```

**Review by atscott — CHANGES_REQUESTED** [review]
AGENT: One remaining detail: in `activate_routes.ts`, `context.injector` is a getter that evaluates to `this.route?.snapshot._environmentInjector ?? this.rootInjector`.

At the moment of detachment, `context.route` still points to the detached route. If that route has route-level `providers: [...]`, `context.injector` will pass the detached route's `_environmentInjector` to the new `ChildrenOutletContexts`.

Instead, `OutletContext` should re-initialize `this.children` with its own `this.rootInjector` (matching its constructor):

In `packages/router/src/router_outlet_context.ts`:
```ts
export class OutletContext {
  // ...
  constructor(private readonly rootInjector: EnvironmentInjector) {
    this.children = new ChildrenOutletContexts(this.rootInjector);
  }

  /** @internal */
  resetChildren(): void {
    this.children = new ChildrenOutletContexts(this.rootInjector);
  }
}
```

In `packages/router/src/operators/activate_routes.ts`:
```diff
     if (context && context.outlet) {
       const componentRef = context.outlet.detach();
       const contexts = context.children.contexts;
-      context.children = new ChildrenOutletContexts(context.injector);
+      context.resetChildren();
       this.routeReuseStrategy.store(route.value.snapshot, {componentRef, route, contexts});
     }
```

---

### Test Example

We should add a test verifying that when a detached route with `providers` is destroyed, subsequent sibling child routes are not contaminated with the destroyed route's injector:

```ts
it('should not leak route injectors from detached routes to sibling routes', async () => {
  class Tab1Service {}

  @Component({selector: 'tab1-cmp', template: 'tab1<router-outlet></router-outlet>', standalone: false})
  class Tab1Cmp {}

  @Component({selector: 'tab2-cmp', template: 'tab2<router-outlet></router-outlet>', standalone: false})
  class Tab2Cmp {}

  @Component({selector: 'tab2-inner-cmp', template: 'tab2-inner', standalone: false})
  class Tab2InnerCmp {
    tab1Service = inject(Tab1Service, {optional: true});
  }

  @NgModule({
    declarations: [Root, TabsShellCmp, Tab1Cmp, Tab2Cmp, Tab2InnerCmp],
    imports: [...ROUTER_DIRECTIVES],
    providers: [
      {provide: RouteReuseStrategy, useClass: ReusableStrategy},
      provideRouter([
        {
          path: 'tabs',
          component: TabsShellCmp,
          children: [
            {
              path: 'tab1',
              data: {reusable: true},
              providers: [Tab1Service],
              component: Tab1Cmp,
            },
            {
              path: 'tab2',
              component: Tab2Cmp,
              children: [{path: 'inner', component: Tab2InnerCmp}],
            },
          ],
        },
      ]),
    ],
  })
  class TestModule {}

  TestBed.configureTestingModule({imports: [TestModule]});
  const router = TestBed.inject(Router);
  const strategy = TestBed.inject(RouteReuseStrategy) as ReusableStrategy;
  const fixture = await createRoot(router, Root);

  // 1. Navigate to tab1 (instantiates tab1 with Tab1Service in its route injector)
  router.navigateByUrl('/tabs/tab1');
  await advance(fixture);

  // 2. Switch to tab2 (detaches tab1)
  router.navigateByUrl('/tabs/tab2');
  await advance(fixture);

  // 3. Destroy tab1's handle
  const tab1Route = router.routerState.snapshot.root.firstChild!.children.find(c => c.routeConfig?.path === 'tab1')!;
  const handle = strategy.retrieve(tab1Route)!;
  destroyDetachedRouteHandle(handle);
  strategy.store(tab1Route, null);

  // 4. Navigate into tab2's child route
  router.navigateByUrl('/tabs/tab2/inner');
  await advance(fixture);

  const innerCmp = fixture.debugElement.query(By.directive(Tab2InnerCmp)).componentInstance as Tab2InnerCmp;
  expect(innerCmp).toBeTruthy();
  expect(innerCmp.tab1Service).toBeNull();
});
```

File: packages/router/src/router_outlet_context.ts [hunk 4]
```diff
@@ -28,6 +28,18 @@ export class OutletContext {
...
   }
+
+  /**
+   * Replaces `children` with a fresh, empty `ChildrenOutletContexts`, keeping the same root
+   * injector this context was created with. Used when a route subtree is detached for
+   * `RouteReuseStrategy` so the detached component's stored contexts stay isolated from whatever
+   * activates next in this outlet.
+   *
+   * @internal
+   */
```
Comment (atscott) [inline packages/router/src/router_outlet_context.ts:35]: AGENT: Let's simplify this JSDoc. An `@internal` tag or a simple 1-line description is sufficient here without the extra narrative.

File: packages/router/src/router_outlet_context.ts [hunk 5]
```diff
@@ -37,8 +49,15 @@ export class OutletContext {
...
-  private contexts = new Map<string, OutletContext>();
+  /**
+   * Contexts for child outlets, keyed by outlet name.
+   *
+   * Exposed so that a detached route subtree can be stored together with its child contexts
+   * without wiping this map (see `RouteReuseStrategy` handling in `activate_routes.ts`).
+   *
+   * @internal
+   */
+  contexts = new Map<string, OutletContext>();
```
Comment (atscott) [inline packages/router/src/router_outlet_context.ts:52]: AGENT: We can remove the second paragraph explaining why the property was made public for this PR.

File: packages/router/src/operators/activate_routes.ts [hunk 6]
```diff
@@ -107,7 +107,14 @@ export class ActivateRoutes {
 
     if (context && context.outlet) {
       const componentRef = context.outlet.detach();
-      const contexts = context.children.onOutletDeactivated();
+      // The detached component and its child `RouterOutlet`s stay alive and keep referencing this
+      // `ChildrenOutletContexts` instance. Take its map as-is (unlike the destruction path, which
+      // calls `onOutletDeactivated()` to prune) and give the `OutletContext` a fresh instance, so:
+      //   - the detached component keeps rendering from the stored map once re-attached (#57285);
+      //   - whatever activates next in this same parent outlet (e.g. a sibling tab) can't mutate
+      //     or wipe the detached component's stored map.
```
Comment (atscott) [inline packages/router/src/operators/activate_routes.ts]: AGENT: Let's condense this comment to a single line or remove it:
```ts
// Reset child contexts so subsequent activations in this outlet do not mutate the detached tree.
```

File: packages/router/test/integration/route_reuse_strategy.spec.ts [hunk 7]
```diff
@@ -433,6 +433,268 @@ export function routeReuseIntegrationSuite() {
       expect(createdComps).toEqual(['parent', 'child', 'child']);
     });
 
+    it('should render child routes on reused list when outer shell is destroyed and recreated', async () => {
```
Comment (atscott) [inline packages/router/test/integration/route_reuse_strategy.spec.ts:436]: AGENT: Let's clean up the comments in these tests:
- Trim the multi-line narrative blocks at the top of each `it` block down to just the issue reference.
- Remove the line-by-step commentary (`// Start on an unrelated route...`, `// Go to /events...`, `// Leave again...`, etc.) and inline route config annotations—the test structure and assertions already speak for themselves.

**Review by atscott — APPROVED** [review]

**Comment by JeanMeche** [comment]
Caretaker note: presubmit Failures are pre-existing

**Comment by crisbeto** [comment]
This PR was merged into the repository. The changes were merged into the following branches:

- main: 8227e5cf6d157d6e472f963da68a804813ad26ed
- 22.1.x: 95c01e9cdea251e09d321a01e9b982b2fe1b7873
